### DISCUSSION TITLE
MySQLでNULLが起こす弊害

### DISCUSSION THREAD

**Discussion by yuyakawai0411** [kakeru-one/web-engineering-roundtable#52] [General]
### はじめに
- プログラミング言語では、nullは値として扱っており、比較演算子を使った結果(nullに対して)の真理値はtrue, falseの2値である。
  - console.log(true == null) => false、console.log(12 > null) => エラー
- SQLでは、NULLは値として扱っておらず、比較演算子を使った結果(NULLに対して)の真理値はtrue, false, unkownの3値である。
  - id = NULL => unkown、id > NULL => unkown

#### 1.ANDにNULLが混じるとtrueが返らなくなる
例. a=2, b=5, c=NULLの時、
1. WHERE a < b AND b > c => unkownが返る
  - ANDのパワーバランスは、false > unkonwn > trueのため、true AND unkonwnはunkonwnになる
2. WHERE a > b OR b > c => unkownが返る
  - ORのパワーバランスは、true > unkown > falseのため、false OR unkonwnはunkonwnになる

#### 2.NOT IN のサブクエリで使用されるテーブルに NULL が含まれているとなにも返らなくなる
以下のURLの「NULLと NOT IN述語」を確認
https://mickindex.sakura.ne.jp/database/db_3vl.html

**Comment by yuyakawai0411** [comment]
## 四則演算にNULLが含まれたとき、NULLが伝搬してしまう。
### はじめに
- 加算: どちらかがNULLの場合、結果はNULL
- 減算: どちらかがNULLの場合、結果はNULL
- 乗算: どちらかがNULLの場合、結果はNULL
- 除算: どちらかがNULLの場合、結果はNULL
#### 起きうる弊害
例えば以下のようなSQLがあったとき、taxの入力がない購入品があったとき、その購入品はヒットしなくなる
```
SELECT *
FROM (
    SELECT price * tax AS total_price
    FROM purchases
)
WHERE total_price >= 1000;
```

**Comment by yuyakawai0411** [comment]
## どうやってNULLを撲滅するか？
1. 最小値、最大値を使って埋める
- 数値の場合は、NULLの代わりに0といった最小値を使う
- 日付の場合は、NULLの代わりに9999-12-31といった最大値を使う
2. 未入力か入力済みかの判断を別のカラムに委ねる
例えば金額を保存するときに、NULLの代わりに0で埋めてしまうと、0円なのか未入力なのかの判別ができなくなる。
金額を入力するときに単位(unit_type)の一緒に保存するようにし、金額を入力した場合はamount=0, unit_type='円'で保存、金額を入力していないときはamount=0, unit_type='undefined'で保存すると区別できる

**Comment by yuyakawai0411** [comment]
## 疑問
- SQLでは他のプログラミング言語と異なり、NULLを値であつかっていない理由はなにか？

**Reply by kakeru-one** [reply]
> 通常の列の値と違って、NULL は行のどこかに余分なビットを持つことで実装されている。そのため記憶領域を圧迫したり、検索パフォーマンスを悪化させる。
https://mickindex.sakura.ne.jp/database/db_getout_null.html

上記がよくわからない。
